### DISCUSSION TITLE
Gitlab SSO - Error processing task in ValidateSSOAccessTask

### DISCUSSION THREAD

**Discussion by arunoruto** [outline/outline#6966] [Self Hosting]
When running Outline using our local Gitlab instance as the Auth provider I get the following error in the logs:
```json
{"label":"authentication","level":"debug","message":"Existing token is still valid, skipping refresh"}
{"label":"worker","level":"info","message":"ValidateSSOAccessTask running","userId":"<id>"}
{"error":"All promises were rejected","level":"error","message":"Error processing task in ValidateSSOAccessTask","stack":"AggregateError: All promises were rejected"}
```
A similar problem has been shown in #4152.

**To Reproduce**
I setup Gitlab as my OIDC provider with the following variables:
```
OIDC_AUTH_URI=https://${GITLAB_DOMAIN}/oauth/authorize
OIDC_TOKEN_URI=https://${GITLAB_DOMAIN}/oauth/token
OIDC_USERINFO_URI=https://${GITLAB_DOMAIN}/oauth/userinfo
OIDC_USERNAME_CLAIM=username
OIDC_DISPLAY_NAME=GitLab
OIDC_SCOPES=openid email

OIDC_CLIENT_ID=...
OIDC_CLIENT_SECRET=...
```

**Expected behavior**
Idealy this error should not appear. Maybe it is introduced with some new gitlab updates

**Outline (please complete the following information):**
- Install: self-hosted, docker
- Version: 0.76.1 [sha256:2e67fc8d32a1efcf7df76affe32b5545f8205bf6ce03e84a477a525bf90cb910](https://hub.docker.com/layers/outlinewiki/outline/0.76.1/images/sha256-46870384078687a00e4caf101c3781eafdbae8c70e888a262daa2e5d20f11a4d)

**Desktop (please complete the following information):**
 - OS: NixOS
 - Browser: Chrome
 - Version 125.0.6422.60

**Comment by linear** [comment]
<p><a href="https://linear.app/outline/issue/OLN-396/gitlab-sso-error-processing-task-in-validatessoaccesstask">OLN-396 Gitlab SSO - Error processing task in ValidateSSOAccessTask</a></p>

**Comment by tommoor** [comment]
Which version number are you running? I'm not clear on how to map the sha you provided to a version

**Comment by arunoruto** [comment]
> Which version number are you running? I'm not clear on how to map the sha you provided to a version

Sorry about that, I thought by hash, it meant the hash of the docker image. I am running the `latest` tag, which is currently pointing to version `0.76.1`.

**Comment by tommoor** [comment]
The true error is not visible as it's masked, there is a change here which will expose it in the next release:
https://github.com/outline/outline/commit/3b9cbb08c880097e49e1cb0d3dcf4c1700588f00

**Comment by arunoruto** [comment]
I am not sure if it relevant to the issue, but I am also getting the following error in the Database:
```
2024-05-30 10:55:05.135 UTC [153] ERROR:  column user.isAdmin does not exist at character 52
2024-05-30 10:55:05.135 UTC [153] STATEMENT:  SELECT "user"."id", "user"."email", "user"."name", "user"."isAdmin", "user"."isViewer", "user"."jwtSecret", "user"."lastActiveAt", "user"."lastActiveIp", "user"."lastSignedInAt", "user"."lastSignedInIp", "user"."lastSigninEmailSentAt", "user"."suspendedAt", "user"."flags", "user"."preferences", "user"."notificationSettings", "user"."language", "user"."avatarUrl", "user"."suspendedById", "user"."invitedById", "user"."teamId", "user"."createdAt", "user"."updatedAt", "user"."deletedAt", "team"."id" AS "team.id", "team"."name" AS "team.name", "team"."subdomain" AS "team.subdomain", "team"."domain" AS "team.domain", "team"."defaultCollectionId" AS "team.defaultCollectionId", "team"."avatarUrl" AS "team.avatarUrl", "team"."sharing" AS "team.sharing", "team"."inviteRequired" AS "team.inviteRequired", "team"."signupQueryParams" AS "team.signupQueryParams", "team"."guestSignin" AS "team.guestSignin", "team"."documentEmbeds" AS "team.documentEmbeds", "team"."memberCollectionCreate" AS "team.memberCollectionCreate", "team"."defaultUserRole" AS "team.defaultUserRole", "team"."preferences" AS "team.preferences", "team"."suspendedAt" AS "team.suspendedAt", "team"."lastActiveAt" AS "team.lastActiveAt", "team"."createdAt" AS "team.createdAt", "team"."updatedAt" AS "team.updatedAt", "team"."deletedAt" AS "team.deletedAt" FROM "users" AS "user" INNER JOIN "teams" AS "team" ON "user"."teamId" = "team"."id" AND ("team"."deletedAt" IS NULL) WHERE ("user"."deletedAt" IS NULL AND "user"."id" = '<id>');
```

Yesterday I didn't notice it, sine I was logged in, but now I am even unable to login to my outline instance using Gitlab.

**Comment by tommoor** [comment]
It seems like you went forward a release which migrated the database and then downgraded the Outline version?

**Comment by arunoruto** [comment]
> It seems like you went forward a release which migrated the database and then downgraded the Outline version?

That is true! Sorry, my mistake. I thought that I could roll back a version or two to get on the same state where it worked.

I tried to provision a new OIDC id and secret, and when I try to authorize Gitlab, I get the following lines in the lof:
```
{"error":"connect EHOSTUNREACH <IP>:443","level":"error","message":"Error during authentication","stack":"Error: connect EHOSTUNREACH <IP>:443\n    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1606:16)"}
{"label":"http","level":"info","message":"  --> GET /auth/oidc.callback?code=<code>"}
```
I am running both outline and docker behind a traefik reverse proxy. I am not sure if that could make any problems. But it does resolve the IP correctly.

**Comment by tommoor** [comment]
It is possible to downgrade but you would also need to rollback the database changes or restore from a database backup. There isn't really a clean way to perfectly rollback the database in all cases so it's a manual measure of running `yarn db:rollback` until everything works again, in release v0.75.0 for example there were 11 database migrations that happened.

I would recommend moving to the current software version. I'm going to move this to a discussion.

**Comment by arunoruto** [comment]
Okay, I got finally a solution to the problem! I had a related problem with mattermost recently, where the obtained token wasn't valid, but I did not see a connection to outline until today. I found [this](https://gitlab.com/gitlab-org/omnibus-gitlab/-/issues/2984#note_1950665902) issue regardin mattermost and it seems like there is a problem with the https connection if it is done via a reverse proxy. I solved the problem by pointing all endpoints to a http URL except the auth endpoint, that one still has to point the the normal https external URL.

I tried the same approach with outline, just using the container name of the gitlab instance instead of `localhost`, and it works!
TLDR; the final relevant OIDC changes are:
```env
OIDC_AUTH_URI=https://gitlab.example.com/oauth/authorize
OIDC_TOKEN_URI=http://gitlab/oauth/token
OIDC_USERINFO_URI=http://gitlab/oauth/userinfo
```
where `gitlab.example.com` is your external gitlab URL and `gitlab` in the http URLs is the container name. You can also try a pinging the name from inside the outline container to see if it es reachable!
